docs/source/features/optimizations/activation_recomputation.rst

```diff
@@ -35,4 +35,18 @@ This is because the input sizes of softmax, dropout, and qkv dot-product attenti
 However, their recomputation cost is relatively smaller than the other linear projection layers that are linear with the hidden size square.
 
 Self-attention recomputation is hard-enabled when using FlashAttention, which is supported in Transformer Engine.
-Also, a user can use the self-attention recomputation without FlashAttention by setting ``activations_checkpoint_granularity=selective``.
\ No newline at end of file
+Also, a user can use the self-attention recomputation without FlashAttention by setting ``activations_checkpoint_granularity=selective``.
+
+Scheme of full and selective checkpointing granularity:
+
+.. image:: https://github.com/NVIDIA/NeMo/releases/download/v2.0.0rc0/asset-post-activation-recomputation-exampe-2.jpg
+    :align: center
+    :alt: activation-recomputation-example-2
```

Comment:
The following is a technical edit of activation_recomputation.rst by line numbers. This file was read-only, so I had to add edits here in the comment field.

Line 3. Suggest adding an introductory sentence that explains what activation recomputation is (summarized from arxiv and HF). For example:

Activation recomputation is a technique used in training large neural network models, such as transformer models, to manage the memory constraints of the device. It addresses the challenge of storing large numbers of intermediate activations—outputs from each layer of the network—required during the back-propagation phase for gradient computation.

Line 11 - 13. suggested revision. Please verify change from "NeMo" to "NeMo Framework".

NeMo Framework enables transformer layer recomputation, a technique that checkpoints the input at each transformer layer and then recalculates the activations for subsequent layers. Transformer layer recomputation significantly reduces the activation memory usage. However, this approach also leads to a 30% increase in the computational demand for each transformer layer, attributable to the necessity of reprocessing the full forward computation for the layer.

Line 14-15 suggested revision.

NeMo Framework also supports partial transformer layer recomputation. This feature is beneficial when recomputing a few transformer layers to fit the training workload on GPU memory. By doing so, it eliminates the need to recompute the remaining layers.

Line 17 -18. fix capitalization, use active voice, suggested revision.

You can enable transformer layer recomputation by setting ``activations_checkpoint_granularity=full``. Additionally, you can specify the number of transformer layers to recompute by setting ``activations_checkpoint_num_layers`` along with ``activations_checkpoint_method=block``.

Line 19. fix capitalization, use active voice, suggested revision.

If you set ``activations_checkpoint_num_layers`` as the total number of layers, the inputs of all transformer layers are checkpointed and recomputed.

Line 21. suggested revision.

If the virtual pipelining is used, ``activations_checkpoint_num_layers`` indicates the layers per virtual pipeline stage.

Line 23. fix capitalization, suggested revision.

NeMo Framework also supports checkpointing the input to a block of multiple consecutive transformer layers, effectively making this block the unit of granularity for recomputation.

Line 25. fix capitalization

Thus, it is only beneficial for memory savings when the model has many transformer layers or the intermediate layers of a transformer layer hold relatively small activation stores.

Line 26. fix capitalization, use active voice, suggested revision.

You can enable this recomputation mode by setting ``activations_checkpoint_method=uniform``. Additionally, you can set the number of transformer layers per recomputation block using ``activations_checkpoint_num_layers``.

Line 31 - 35. various corrections, suggested revision.

NeMo Framework supports the self-attention recomputation that checkpoints the inputs of each self-attention block and recomputes the intermediate input activations. This is a cost-efficient recomputation method that achieves higher memory savings with lower recomputation cost.

The intermediate layers within the self-attention block are responsible for the majority of activation memory usage. 
This is because the input sizes of softmax, dropout, and qkv dot-product attention layers have a memory complexity that is proportional to the square of the sequence length. 

However, the recomputation cost of these layers is relatively smaller compared to that of the other linear projection layers, which scale linearly with the square of the hidden size.

Line 37. suggested revision.

Self-attention recomputation is hard-enabled when using Flash Attention, which the Transformer Engine supports. Additionally, you can opt for self-attention recomputation without Flash Attention by configuring ``activations_checkpoint_granularity=selective``.